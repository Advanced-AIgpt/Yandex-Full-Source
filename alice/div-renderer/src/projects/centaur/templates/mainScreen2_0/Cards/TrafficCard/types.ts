import { IAbstractCardProps } from '../types';

export interface ITrafficForecast {
    time: number | null;
    trafficValue?: number | null;
}

export interface ITrafficCardProps extends IAbstractCardProps {
    type: 'traffic';
    city?: string | null;
    text?: string | null;
    trafficValue?: number | null;
    forecasts?: ITrafficForecast[];
}
