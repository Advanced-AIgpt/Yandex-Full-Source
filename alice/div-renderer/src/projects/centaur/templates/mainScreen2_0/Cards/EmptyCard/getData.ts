import { IEmptyCardProps } from './types';
import { ICardDataAdapter } from '../types';

export const getEmptyCardData: ICardDataAdapter<IEmptyCardProps> = function getEmptyCardData(card, requestState) {
    if (card && card.VacantCardData) {
        return {
            type: 'empty',
            requestState,
        };
    }
    return null;
};
