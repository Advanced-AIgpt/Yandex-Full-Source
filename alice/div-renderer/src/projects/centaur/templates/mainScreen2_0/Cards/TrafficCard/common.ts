import { ITrafficForecast } from './types';

export function sortTimeOfDay(originalArray: Readonly<ITrafficForecast[]>): ITrafficForecast[] {
    const forecastsFirst = [];
    const forecastsLast = [];

    for (const forecastElement of originalArray) {
        if (forecastElement.time === null || typeof forecastElement.time === 'undefined') {
            continue;
        }

        if (forecastsFirst.length === 0) {
            forecastsFirst.push(forecastElement);
        } else if (forecastsLast.length > 0) {
            forecastsLast.push(forecastElement);
        } else if (forecastElement.time - (forecastsFirst[forecastsFirst.length - 1].time as number) === 1) {
            forecastsFirst.push(forecastElement);
        } else {
            forecastsLast.push(forecastElement);
        }
    }

    return [...forecastsLast, ...forecastsFirst];
}
