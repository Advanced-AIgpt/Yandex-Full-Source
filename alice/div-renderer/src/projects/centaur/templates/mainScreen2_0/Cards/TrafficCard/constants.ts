import { colorWhite } from '../../../../style/constants';

export const colorTrafficFree = '#66B616';
export const colorTrafficLight = '#FAE268';
export const colorTextTrafficLight = '#785C00';
export const colorTrafficHard = '#E55245';
export const colorTrafficNA = '#CCCCCC';
export const colorTrafficOff = '#FFFFFF';

export const CURRENT_TRAFFIC_VALUE_SIZE = 68;
export const CURRENT_TRAFFIC_SPACE_BETWEEN_ELEMENTS = 22;
export const TRAFFIC_FORECAST_ITEM_WIDTH = 52;
export const TRAFFIC_FORECAST_ITEM_HEIGHT = 36;
export const TRAFFIC_FORECAST_BORDER_RADIUS = Math.floor(
    Math.min(
        TRAFFIC_FORECAST_ITEM_WIDTH,
        TRAFFIC_FORECAST_ITEM_HEIGHT,
    ) / 2,
);

const TRAFFIC_COLOR_SET_FREE = [colorTrafficFree, colorWhite] as const;
const TRAFFIC_COLOR_SET_LIGHT = [colorTrafficLight, colorTextTrafficLight] as const;
const TRAFFIC_COLOR_SET_HARD = [colorTrafficHard, colorWhite] as const;

export const TRAFFIC_COLOR_SET: Readonly<Readonly<[string, string]>[]> = [
    TRAFFIC_COLOR_SET_FREE, //     0
    TRAFFIC_COLOR_SET_FREE, //     1
    TRAFFIC_COLOR_SET_FREE, //     2
    TRAFFIC_COLOR_SET_FREE, //     3
    TRAFFIC_COLOR_SET_LIGHT, //    4
    TRAFFIC_COLOR_SET_LIGHT, //    5
    TRAFFIC_COLOR_SET_HARD, //     6
    TRAFFIC_COLOR_SET_HARD, //     7
    TRAFFIC_COLOR_SET_HARD, //     8
    TRAFFIC_COLOR_SET_HARD, //     9
    TRAFFIC_COLOR_SET_HARD, //     10
];
