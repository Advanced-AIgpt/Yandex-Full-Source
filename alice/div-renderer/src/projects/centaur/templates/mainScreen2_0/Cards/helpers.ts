export function getCardMainScreenId({ rowIndex, colIndex }: { rowIndex?: number; colIndex?: number }) {
    return `main_screen_card_${rowIndex || 0}_${colIndex || 0}`;
}
