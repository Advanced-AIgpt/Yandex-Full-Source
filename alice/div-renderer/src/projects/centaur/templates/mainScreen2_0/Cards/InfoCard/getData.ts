import { sample } from 'lodash';
import { IInfoCardProps } from './types';
import { DISCOVERY_CARDS } from '../../../mainScreen/discovery';
import { ICardDataAdapter } from '../types';
import { textAction } from '../../../../../../common/actions';

export const getInfoCardData: ICardDataAdapter<IInfoCardProps> = function getInfoCardData(card, requestState) {
    if (typeof card.InfoCardData !== 'undefined' && card.InfoCardData !== null) {
        const data = card.InfoCardData.Title === 'test' ? DISCOVERY_CARDS[0] : sample(DISCOVERY_CARDS) || DISCOVERY_CARDS[0];

        return {
            type: 'info',
            title: data.title,
            image_background: data.bg,
            description: data.description,
            requestState,
            actions: [
                {
                    log_id: 'discovery.mainscreen.execute',
                    url: textAction(data.description),
                },
            ],
        };
    }
    return null;
};
