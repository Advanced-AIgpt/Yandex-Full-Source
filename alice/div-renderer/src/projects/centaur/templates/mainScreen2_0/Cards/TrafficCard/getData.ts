import { compact } from 'lodash';
import { ITrafficCardProps, ITrafficForecast } from './types';
import { NAlice } from '../../../../../../protos';
import { sortTimeOfDay } from './common';
import { ICardDataAdapter } from '../types';

function forecastDataAdapter(forecastItem: NAlice.NData.ITTrafficForecastData | null | undefined): ITrafficForecast | null {
    if (!forecastItem || forecastItem.Hour === null || typeof forecastItem.Hour === 'undefined') {
        return null;
    }

    return {
        time: forecastItem.Hour,
        trafficValue: forecastItem.Score,
    };
}

export const getTrafficCardData: ICardDataAdapter<ITrafficCardProps> =
    function getTrafficCardData(card, requestState) {
        if (typeof card.TrafficCardData !== 'undefined' && card.TrafficCardData !== null) {
            const forecasts = compact(card.TrafficCardData.Forecast?.map(forecastDataAdapter));
            const sortedForecasts = sortTimeOfDay(forecasts).slice(0, 4);

            return {
                type: 'traffic',
                text: card.TrafficCardData.Message,
                trafficValue: card.TrafficCardData.Score,
                city: card.TrafficCardData.City,
                requestState,
                forecasts: sortedForecasts,
            };
        }
        return null;
    };
