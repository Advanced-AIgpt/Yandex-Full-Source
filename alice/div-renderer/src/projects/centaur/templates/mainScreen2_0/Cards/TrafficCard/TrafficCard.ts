import {
    ContainerBlock,
    Div,
    FixedSize,
    ImageBlock,
    TextBlock,
    WrapContentSize,
    SolidBackground,
    MatchParentSize,
} from 'divcard2';
import { compact } from 'lodash';
import {
    ITrafficCardProps,
    ITrafficForecast,
} from './types';
import {
    text28m,
    title88m,
} from '../../../../style/Text/Text';
import {
    colorWhite,
    colorWhiteOpacity50,
} from '../../../../style/constants';
import { BasicTextCard } from '../BasicTextCard';
import BasicErrorCard from '../BasicErrorCard';
import { logger } from '../../../../../../common/logger';
import { getS3Asset } from '../../../../helpers/assets';
import EmptyDiv from '../../../../components/EmptyDiv';

import {
    CURRENT_TRAFFIC_VALUE_SIZE,
    CURRENT_TRAFFIC_SPACE_BETWEEN_ELEMENTS,
    TRAFFIC_FORECAST_BORDER_RADIUS,
    TRAFFIC_FORECAST_ITEM_HEIGHT, TRAFFIC_COLOR_SET, colorTrafficNA, TRAFFIC_FORECAST_ITEM_WIDTH,
} from './constants';
import getColorSet, { IColorSet } from '../../../../style/colorSet';
import { getCardMainScreenId } from '../helpers';

function getStateColor(trafficScore: number | null | undefined) : string {
    return (
        trafficScore !== null && typeof trafficScore !== 'undefined' && TRAFFIC_COLOR_SET[trafficScore][0]
    ) || colorTrafficNA;
}

function CurrentTraffic({
    text,
    trafficValue,
}: {
    text?: string;
    trafficValue: number;
}): Div {
    return new ContainerBlock({
        orientation: 'horizontal',
        content_alignment_vertical: 'center',
        content_alignment_horizontal: 'left',
        items: compact([
            new TextBlock({
                ...title88m,
                text_color: colorWhite,
                width: new WrapContentSize(),
                text: `${trafficValue}`,
            }),
            TrafficLights({
                size: CURRENT_TRAFFIC_VALUE_SIZE,
                color: getStateColor(trafficValue),
            }),
            text && new TextBlock({
                ...text28m,
                margins: {
                    left: CURRENT_TRAFFIC_SPACE_BETWEEN_ELEMENTS,
                },
                text_color: colorWhiteOpacity50,
                text_alignment_horizontal: 'left',
                width: new MatchParentSize(),
                text,
            }),
        ]),
    });
}

function TrafficLights({
    color,
    size,
}: {
    color: string;
    size: number;
}) {
    return new ContainerBlock({
        orientation: 'overlap',
        content_alignment_vertical: 'top',
        content_alignment_horizontal: 'center',
        width: new FixedSize({ value: size }),
        height: new FixedSize({ value: size }),
        margins: {
            left: CURRENT_TRAFFIC_SPACE_BETWEEN_ELEMENTS,
        },
        items: [
            new ContainerBlock({
                orientation: 'horizontal',
                content_alignment_vertical: 'center',
                content_alignment_horizontal: 'center',
                margins: {
                    top: 1,
                },
                items: [
                    new EmptyDiv({
                        width: new FixedSize({ value: size * 0.06 }),
                    }),
                    new ContainerBlock({
                        width: new FixedSize({ value: size * 0.94 }),
                        height: new FixedSize({ value: size * 0.94 }),
                        border: {
                            corner_radius: Math.floor(size / 2),
                        },
                        background: [
                            new SolidBackground({ color }),
                        ],
                        items: [],
                    }),
                ],
            }),
            new ImageBlock({
                image_url: getS3Asset('main-screen/cards/jams/traffic_lights_icon.png'),
                alignment_vertical: 'top',
                alignment_horizontal: 'right',
                preload_required: 1,
            }),
        ],
    });
}

function getTrafficColors(value: number | null | undefined): Readonly<[string, string]> {
    if (value === null || typeof value === 'undefined' || typeof TRAFFIC_COLOR_SET[value] === 'undefined') {
        return [colorTrafficNA, colorWhite];
    }
    return TRAFFIC_COLOR_SET[value];
}

function TrafficForecastItem({ time, trafficValue }: ITrafficForecast, colorSet: IColorSet) {
    const [bgColor, color] = getTrafficColors(trafficValue);

    return new ContainerBlock({
        orientation: 'vertical',
        width: new WrapContentSize(),
        content_alignment_vertical: 'center',
        content_alignment_horizontal: 'center',
        items: [
            new TextBlock({
                ...text28m,
                width: new FixedSize({ value: TRAFFIC_FORECAST_ITEM_WIDTH }),
                height: new FixedSize({ value: TRAFFIC_FORECAST_ITEM_HEIGHT }),
                text_alignment_horizontal: 'center',
                text_alignment_vertical: 'center',
                text_color: color,
                border: {
                    corner_radius: TRAFFIC_FORECAST_BORDER_RADIUS,
                },
                background: [
                    new SolidBackground({ color: bgColor }),
                ],
                text: `${trafficValue}`,
            }),
            new TextBlock({
                ...text28m,
                margins: {
                    top: 4,
                },
                width: new MatchParentSize(),
                height: new FixedSize({ value: TRAFFIC_FORECAST_ITEM_HEIGHT }),
                text_alignment_horizontal: 'center',
                text_color: colorSet.textColorOpacity50,
                text: `${time}`,
            }),
        ],
    });
}

function TrafficForecast({ forecasts, colorSet }: {
    forecasts: ITrafficForecast[];
    colorSet: IColorSet;
}): Div {
    const forecastItems: Div[] = [];

    for (let i = 0; i < forecasts.length; i++) {
        forecastItems.push(TrafficForecastItem(forecasts[i], colorSet));
        if (i + 1 < forecasts.length) {
            forecastItems.push(new EmptyDiv({
                width: new MatchParentSize({ weight: 1 }),
            }));
        }
    }

    return new ContainerBlock({
        orientation: 'horizontal',
        content_alignment_vertical: 'center',
        content_alignment_horizontal: 'left',
        width: new MatchParentSize(),
        items: forecastItems,
    });
}

export default function TrafficCard({
    trafficValue,
    forecasts,
    actions,
    longtap_actions,
    city,
    rowIndex,
    colIndex,
}: Partial<ITrafficCardProps>): Div {
    const colorSet = getColorSet();

    if (trafficValue === null || typeof trafficValue === 'undefined' || !forecasts) {
        logger.error('The traffic card should have a text, trafficValue, trafficColor and forecasts, but some of this was not transmitted');

        return BasicErrorCard({
            colIndex,
            rowIndex,
            title: 'Пробки',
            description: 'Ошибка сети :(\nНе могу показать уровень пробок. Проверьте интернет.',
            actions,
            longtap_actions,
        });
    }

    const cardData: Parameters<typeof BasicTextCard>[0] = {
        id: getCardMainScreenId({ colIndex, rowIndex }),
        title: city || 'Пробки',
        items: [
            CurrentTraffic({ trafficValue }),
            new EmptyDiv({
                height: new MatchParentSize(),
            }),
            TrafficForecast({ forecasts, colorSet }),
        ],
        actions,
        longtap_actions,
        colorSet,
    };

    return BasicTextCard(cardData);
}
