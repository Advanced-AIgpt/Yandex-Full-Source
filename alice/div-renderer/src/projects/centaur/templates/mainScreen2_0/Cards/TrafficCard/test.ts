import { AnonymizeDataForSnapshot } from '../../../../../../common/helpers/dev';
import TrafficCard from './TrafficCard';
import { NAlice } from '../../../../../../protos';
import { getTrafficCardData } from './getData';
import { logger } from '../../../../../../common/logger';

const TCardData = NAlice.NData.TCentaurMainScreenMyScreenData.TColumn.TCardData;

jest.mock('../../../../../../common/logger');

describe('Main screen empty card', () => {
    it('should get actual data', () => {
        expect(getTrafficCardData(new TCardData({}))).toBeNull();

        expect(getTrafficCardData(new TCardData({
            YouTubeCardData: {},
        }))).toBeNull();

        expect(getTrafficCardData(new TCardData({
            YouTubeCardData: {},
            TrafficCardData: {
            },
        }))).not.toBeNull();

        expect(getTrafficCardData(new TCardData({
            TrafficCardData: {
                Forecast: [],
                City: 'Москва',
                Level: 'level',
                Score: 7,
                Message: 'Облачно, без пробок',
            },
        }))).not.toBeNull();
    });

    it('should match snapshot', () => {
        expect(AnonymizeDataForSnapshot(TrafficCard({
            type: 'traffic',
            text: 'Текст о пробках',
            city: 'Москва',
            trafficValue: 7,
            forecasts: [
                {
                    time: 17,
                    trafficValue: 8,
                },
            ],
        }))).toMatchSnapshot();

        expect(logger.error).toHaveBeenCalledTimes(0);
    });

    it('should match snapshot with error', () => {
        expect(AnonymizeDataForSnapshot(TrafficCard({
            type: 'traffic',
        }))).toMatchSnapshot();

        expect(logger.error).toHaveBeenCalled();
    });
});
