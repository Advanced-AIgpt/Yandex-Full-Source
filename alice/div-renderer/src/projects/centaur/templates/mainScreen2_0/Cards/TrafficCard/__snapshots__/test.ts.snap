// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Main screen empty card should match snapshot 1`] = `
Object {
  "background": Array [
    Object {
      "color": "#1aeaf6ec",
      "type": "solid",
    },
  ],
  "border": Object {
    "corner_radius": 28,
  },
  "height": Object {
    "type": "match_parent",
    "weight": 1,
  },
  "id": "main_screen_card_0_0",
  "items": Array [
    Object {
      "items": Array [
        Object {
          "font_size": 32,
          "font_weight": "medium",
          "line_height": 40,
          "text": "Москва",
          "text_color": "#80ffffff",
          "type": "text",
        },
      ],
      "orientation": "horizontal",
      "type": "container",
    },
    Object {
      "delimiter_style": Object {
        "color": "#0fff",
      },
      "height": Object {
        "type": "match_parent",
      },
      "type": "separator",
    },
    Object {
      "content_alignment_horizontal": "left",
      "content_alignment_vertical": "center",
      "items": Array [
        Object {
          "font_size": 88,
          "font_weight": "medium",
          "line_height": 101,
          "text": "7",
          "text_color": "#fff",
          "type": "text",
          "width": Object {
            "type": "wrap_content",
          },
        },
        Object {
          "content_alignment_horizontal": "center",
          "content_alignment_vertical": "top",
          "height": Object {
            "type": "fixed",
            "value": 68,
          },
          "items": Array [
            Object {
              "content_alignment_horizontal": "center",
              "content_alignment_vertical": "center",
              "items": Array [
                Object {
                  "delimiter_style": Object {
                    "color": "#0fff",
                  },
                  "type": "separator",
                  "width": Object {
                    "type": "fixed",
                    "value": 4.08,
                  },
                },
                Object {
                  "background": Array [
                    Object {
                      "color": "#E55245",
                      "type": "solid",
                    },
                  ],
                  "border": Object {
                    "corner_radius": 34,
                  },
                  "height": Object {
                    "type": "fixed",
                    "value": 63.919999999999995,
                  },
                  "items": Array [],
                  "type": "container",
                  "width": Object {
                    "type": "fixed",
                    "value": 63.919999999999995,
                  },
                },
              ],
              "margins": Object {
                "top": 1,
              },
              "orientation": "horizontal",
              "type": "container",
            },
            Object {
              "alignment_horizontal": "right",
              "alignment_vertical": "top",
              "image_url": "https://dialogs.s3.yandex.net/smart_displays/main-screen/cards/jams/traffic_lights_icon.png",
              "preload_required": 1,
              "type": "image",
            },
          ],
          "margins": Object {
            "left": 22,
          },
          "orientation": "overlap",
          "type": "container",
          "width": Object {
            "type": "fixed",
            "value": 68,
          },
        },
      ],
      "orientation": "horizontal",
      "type": "container",
    },
    Object {
      "delimiter_style": Object {
        "color": "#0fff",
      },
      "height": Object {
        "type": "match_parent",
      },
      "type": "separator",
    },
    Object {
      "content_alignment_horizontal": "left",
      "content_alignment_vertical": "center",
      "items": Array [
        Object {
          "content_alignment_horizontal": "center",
          "content_alignment_vertical": "center",
          "items": Array [
            Object {
              "background": Array [
                Object {
                  "color": "#E55245",
                  "type": "solid",
                },
              ],
              "border": Object {
                "corner_radius": 18,
              },
              "font_size": 28,
              "font_weight": "medium",
              "height": Object {
                "type": "fixed",
                "value": 36,
              },
              "line_height": 36,
              "text": "8",
              "text_alignment_horizontal": "center",
              "text_alignment_vertical": "center",
              "text_color": "#fff",
              "type": "text",
              "width": Object {
                "type": "fixed",
                "value": 52,
              },
            },
            Object {
              "font_size": 28,
              "font_weight": "medium",
              "height": Object {
                "type": "fixed",
                "value": 36,
              },
              "line_height": 36,
              "margins": Object {
                "top": 4,
              },
              "text": "17",
              "text_alignment_horizontal": "center",
              "text_color": "#80eaf6ec",
              "type": "text",
              "width": Object {
                "type": "match_parent",
              },
            },
          ],
          "orientation": "vertical",
          "type": "container",
          "width": Object {
            "type": "wrap_content",
          },
        },
      ],
      "orientation": "horizontal",
      "type": "container",
      "width": Object {
        "type": "match_parent",
      },
    },
  ],
  "paddings": Object {
    "bottom": 24,
    "left": 24,
    "right": 24,
    "top": 24,
  },
  "type": "container",
  "width": Object {
    "type": "match_parent",
    "weight": 1,
  },
}
`;

exports[`Main screen empty card should match snapshot with error 1`] = `
Object {
  "background": Array [
    Object {
      "color": "#1aeaf6ec",
      "type": "solid",
    },
  ],
  "border": Object {
    "corner_radius": 28,
  },
  "height": Object {
    "type": "match_parent",
    "weight": 1,
  },
  "id": "main_screen_card_0_0",
  "items": Array [
    Object {
      "items": Array [
        Object {
          "font_size": 32,
          "font_weight": "medium",
          "line_height": 40,
          "text": "Пробки",
          "text_color": "#80ffffff",
          "type": "text",
        },
      ],
      "orientation": "horizontal",
      "type": "container",
    },
    Object {
      "delimiter_style": Object {
        "color": "#0fff",
      },
      "height": Object {
        "type": "match_parent",
      },
      "type": "separator",
    },
    Object {
      "font_size": 28,
      "font_weight": "medium",
      "line_height": 36,
      "text": "Ошибка сети :(
Не могу показать уровень пробок. Проверьте интернет.",
      "text_color": "#80ffffff",
      "type": "text",
    },
  ],
  "paddings": Object {
    "bottom": 24,
    "left": 24,
    "right": 24,
    "top": 24,
  },
  "type": "container",
  "width": Object {
    "type": "match_parent",
    "weight": 1,
  },
}
`;
